### PR TITLE
Add Arcee model support

### PR DESCRIPTION
**Pull request by Crystalcareai**
## Summary

This PR adds support for the Arcee model architecture, laying the groundwork for the upcoming Arcee Foundation Model (AFM) release. Arcee is a decoder-only transformer model based on the Llama architecture with a key modification: it uses ReLU² (ReLU-squared) activation in the MLP blocks instead of SiLU, following recent research showing improved training efficiency with squared activations.

## Model Description

Arcee is architecturally similar to Llama but with the following distinctions:

- **ReLU² activation**: Uses `x * relu(x)` in MLP layers for improved gradient flow
- **Optimized for efficiency**: Designed with training and inference efficiency in mind
- **Extended context**: Supports extended context with RoPE scaling

## Implementation Details

- Modular implementation inheriting from Llama components where applicable
- Custom ArceeMLP class implementing the ReLU² activation
- Full support for all standard transformers features:
  - Flash Attention 2, SDPA, and other attention backends
  - Gradient checkpointing
  - Quantization support (including quantized caches)
  - All standard model variants (CausalLM, SequenceClassification, QuestionAnswering, TokenClassification)

## Testing

- Added comprehensive test suite following standard transformers test patterns
- Tests for all model variants and core functionality
- Specific test for ReLU² activation verification
- RoPE scaling tests including YARN support
- Tested model forward and backward passes
- Verified compatibility with existing architecture
- Model loading and forward passes verified
- Compatibility with existing infrastructure confirmed

### CODE REVIEW

**Comment by Rocketknight1** [comment]
looks good @Crystalcareai! Feel free to ping us whenever you're ready for review. You can also resolve the code style errors with `pip install -e .[quality]` followed by `make style` or `make fixup`

**Comment by Crystalcareai** [comment]
@Rocketknight1 Hey I think I'm ready for a review, Got a lot of the tests passing though still getting some failures that don't seem to be related to my code. Let me know how best I can get this ready for merging.

File: docs/source/en/model_doc/arcee.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
+<!--Copyright 2024 The HuggingFace Team. All rights reserved.
```
Comment (Cyrilvallez) [inline docs/source/en/model_doc/arcee.md]: It's 2025! 🤗

File: src/transformers/models/arcee/__init__.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2022 EleutherAI and the HuggingFace Inc. team. All rights reserved.
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/__init__.py]: wrong dates and companies here as well

File: src/transformers/models/arcee/__init__.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#
+# This code is based on EleutherAI's GPT-NeoX library and the GPT-NeoX
+# and OPT implementations in this library. It has been modified from its
+# original forms to accommodate minor architectural differences compared
+# to GPT-NeoX and OPT used by the Meta AI team that trained the model.
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/__init__.py]: Not true, to remove

File: src/transformers/models/arcee/modular_arcee.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
+# coding=utf-8
+# Copyright 2022 EleutherAI and the HuggingFace Inc. team. All rights reserved.
+#
+# This code is based on EleutherAI's GPT-NeoX library and the GPT-NeoX
+# and OPT implementations in this library. It has been modified from its
+# original forms to accommodate minor architectural differences compared
+# to GPT-NeoX and OPT used by the Meta AI team that trained the model.
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: same comment as above here!

File: src/transformers/models/arcee/modular_arcee.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        self.config = config
+        self.hidden_size = config.hidden_size
+        self.intermediate_size = config.intermediate_size
+        self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=config.mlp_bias)
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=config.mlp_bias)
+        self.act_fn = ACT2FN[config.hidden_act]
+
+    def forward(self, x):
+        down_proj = self.down_proj(self.act_fn(self.up_proj(x)))
+        return down_proj
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: You could actually inherit that directly from Nemotron as it's 1:1 similar!

File: src/transformers/models/arcee/modular_arcee.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        # BC: if there is a 'type' field, copy it to 'rope_type'.
+        if self.rope_scaling is not None and "type" in self.rope_scaling:
+            self.rope_scaling["rope_type"] = self.rope_scaling["type"]
+        rope_config_validation(self)
+
+
+class ArceeRMSNorm(LlamaRMSNorm):
+    """ArceeRMSNorm is identical to LlamaRMSNorm"""
+
+    pass
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: Not strictly needed, will be picked up automatically by other modules!

File: src/transformers/models/arcee/modular_arcee.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+    """
+
+    def __init__(self, config: ArceeConfig):
+        super().__init__(config)
+        # The parent init handles most setup, we just need to ensure our decoder layers are used
+        self.layers = nn.ModuleList(
+            [ArceeDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
+        # Initialize weights and apply final processing
+        self.post_init()
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: Here we can remove all of those - they will be picked up automatically when inheriting the CausalLM class! 🤗

File: src/transformers/models/arcee/modular_arcee.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+class ArceeForCausalLM(LlamaForCausalLM):
+    """Arcee Model transformer with a language modeling head on top (linear layer with weights tied to the input embeddings)."""
+
+    def __init__(self, config):
+        # We need to set config_class before calling super().__init__
+        self.config_class = ArceeConfig
+        super().__init__(config)
+        self.model = ArceeModel(config)
+        # Initialize weights and apply final processing
+        self.post_init()
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: This should be sufficient!

```suggestion
class ArceeForCausalLM(LlamaForCausalLM):
    pass
```

File: src/transformers/models/arcee/modular_arcee.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+    """
+    The Arcee Model transformer with a token classification head on top.
+    """
+
+    def __init__(self, config):
+        self.config_class = ArceeConfig
+        super().__init__(config)
+        self.model = ArceeModel(config)
+        # Initialize weights and apply final processing
+        self.post_init()
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: Similarly for those, we don't actually need to rewrite the __init__ at all 🤗

File: src/transformers/utils/fx.py [hunk 10]
```diff
@@ -116,6 +116,7 @@ def _generate_supported_model_class_names(
 
 _REGULAR_SUPPORTED_MODEL_NAMES_AND_TASKS = [
     "altclip",
+    "arcee",
```
Comment (Cyrilvallez) [inline src/transformers/utils/fx.py]: This change should be reverted - this is more for legacy purposes

File: tests/models/arcee/test_modeling_arcee.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+            "attention_factor": 1.0,
+            "beta_fast": 32,
+            "beta_slow": 1,
+        }
+        model = ArceeModel(config)
+        model.to(torch_device)
+        model.eval()
+        input_ids = torch.randint(0, config.vocab_size, (1, 10)).to(torch_device)
+        with torch.no_grad():
+            model(input_ids)
```
Comment (Cyrilvallez) [inline tests/models/arcee/test_modeling_arcee.py]: Those tests are not needed and can be removed

File: tests/models/arcee/test_modeling_arcee.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+            mlp = layer.mlp
+            # Test with a simple input
+            x = torch.randn(1, 10, config.hidden_size)
+            up_output = mlp.up_proj(x)
+
+            # Verify ReLU² activation: x * relu(x)
+            expected_activation = up_output * torch.relu(up_output)
+            actual_activation = mlp.act_fn(up_output)
+
+            self.assertTrue(torch.allclose(expected_activation, actual_activation, atol=1e-5))
```
Comment (Cyrilvallez) [inline tests/models/arcee/test_modeling_arcee.py]: I don't mind testing this to make sure, but let's not use a for loop if we don't actually use the loop!

File: tests/models/arcee/test_modeling_arcee.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
+# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```
Comment (Cyrilvallez) [inline tests/models/arcee/test_modeling_arcee.py]: date here as well haha

**Review by Cyrilvallez — COMMENTED** [review]
Hey! Very clean first implementation with modular, congrats!! 🤗 We can still make it even simpler though, see my comments 🚀 Also, let's make sure the copyright at the top of files have the correct informations (dates and company names mostly)

But very nice work in general! 🤗

**Comment by pranav4501** [comment]
Hi @Cyrilvallez ,
Thanks for the feedback, made the requested refactoring changes.
Also, while removing the init from the modular implementation as suggested, the generated modeling code does not have self.config_class = ArceeConfig from the previous version. Is that redundant as well?

**Comment by Cyrilvallez** [comment]
> Also, while removing the init from the modular implementation as suggested, the generated modeling code does not have self.config_class = ArceeConfig from the previous version. Is that redundant as well?

Yes, it's already in the PreTrainedModel!

File: src/transformers/models/arcee/modular_arcee.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+            Padding token id.
+        bos_token_id (`int`, *optional*, defaults to 128000):
+            Beginning of stream token id.
+        eos_token_id (`int`, *optional*, defaults to 128001):
+            End of stream token id.
+        pretraining_tp (`int`, *optional*, defaults to 1):
+            Experimental feature. Tensor parallelism rank used during pretraining. Please refer to [this
+            document](https://huggingface.co/docs/transformers/main/perf_train_gpu_many#tensor-parallelism) to
+            understand more about it. This value is necessary to ensure exact reproducibility of the pretraining
+            results. Please refer to [this issue](https://github.com/pytorch/pytorch/issues/76232).
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: This arg should be removed!

File: src/transformers/models/arcee/modular_arcee.py [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    >>> configuration = ArceeConfig()
+
+    >>> # Initializing a model from the AFM-4.5B-Base style configuration
+    >>> model = ArceeModel(configuration)
+
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
+
+    model_type = "arcee"
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py:132]: Here you should add the following TP plan as class attribute (we need to change as the MLP is slightly different from Llama):

```python
base_model_tp_plan = {
        "layers.*.self_attn.q_proj": "colwise",
        "layers.*.self_attn.k_proj": "colwise",
        "layers.*.self_attn.v_proj": "colwise",
        "layers.*.self_attn.o_proj": "rowwise",
        "layers.*.mlp.up_proj": "colwise",
        "layers.*.mlp.down_proj": "rowwise",
    }
```

File: src/transformers/models/arcee/modular_arcee.py [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+        num_key_value_heads=None,
+        hidden_act="relu2",
+        max_position_embeddings=4096,
+        initializer_range=0.02,
+        rms_norm_eps=1e-5,
+        use_cache=True,
+        pad_token_id=None,
+        bos_token_id=128000,
+        eos_token_id=128001,
+        pretraining_tp=1,
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: pretraining_tp to remove here

File: src/transformers/models/arcee/modular_arcee.py [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+            head_dim=head_dim,
+            **kwargs,
+        )
+
+        # Validate the correctness of rotary position embeddings parameters using Arcee's custom validation
+        # BC: if there is a 'type' field, copy it to 'rope_type'.
+        if self.rope_scaling is not None and "type" in self.rope_scaling:
+            self.rope_scaling["rope_type"] = self.rope_scaling["type"]
+        rope_config_validation(self)
+
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: No need to overwrite the check, but we should delete the pretraining_tp attribute to make it disppear in actual config
```suggestion
        del self.pretraining_tp

```

File: src/transformers/models/arcee/modular_arcee.py [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+            num_key_value_heads=num_key_value_heads,
+            hidden_act=hidden_act,
+            max_position_embeddings=max_position_embeddings,
+            initializer_range=initializer_range,
+            rms_norm_eps=rms_norm_eps,
+            use_cache=use_cache,
+            pad_token_id=pad_token_id,
+            bos_token_id=bos_token_id,
+            eos_token_id=eos_token_id,
+            pretraining_tp=pretraining_tp,
```
Comment (Cyrilvallez) [inline src/transformers/models/arcee/modular_arcee.py]: pretraining_tp to remove here as well

File: tests/models/arcee/test_modeling_arcee.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    @slow
+    def test_model_from_pretrained(self):
+        # This test would be enabled once a pretrained model is available
+        # For now, we just test that the model can be instantiated
+        config = ArceeConfig()
+        model = ArceeForCausalLM(config)
+        self.assertIsInstance(model, ArceeForCausalLM)
```
Comment (Cyrilvallez) [inline tests/models/arcee/test_modeling_arcee.py:125]: Would be nice to add a few Integration test based on real checkpoints as well if possible! 🤗 Otherwise you can open another PR later if more convenient for you

**Review by Cyrilvallez — COMMENTED** [review]
Alright, it's perfect! 🤗🤗 Just added some minor comments about the config (most notably, the tp_plan should reflect the new MLP (this is very important for your model to be available as backend in vllm/TGI and other frameworks), but otherwise all good! Great work! I'll merge as soon as you make those small changes 🤗

**Comment by pranav4501** [comment]
@Cyrilvallez Thanks for the feedback, removed the pretraining TP from the configurations and added scaffolding for generation integration testing. We will add more robust integration tests and update the checkpoints with the release.

**Review by Cyrilvallez — APPROVED** [review]
All right, merging! Thanks a lot! TP plan is still wrong, but I'll update it myself after the merge! 🤗🚀
